master/buildbot/process/builder.py

```diff
@@ -436,7 +436,6 @@ def run_cleanups():
         # http://trac.buildbot.net/ticket/2428).
         d = defer.maybeDeferred(build.startBuild,
                                 bs, workerforbuilder)
-        d.addCallback(lambda _: self.buildFinished(build, workerforbuilder))
```

Comment:
Since `Build.startBuild` returns before the build is finished, move this call into `Build.buildFinished